moveit_core/collision_detection/src/collision_env.cpp

```diff
@@ -0,0 +1,298 @@
+    return false;
+  }
+  if (scale > std::numeric_limits<double>::max())
+  {
+    ROS_ERROR_NAMED("collision_detection", "Scale must be finite");
+    return false;
+  }
+  return true;
+}
+
+static inline bool validatePadding(double padding)
+{
+  if (padding < 0.0)
+  {
+    ROS_ERROR_NAMED("collision_detection", "Padding cannot be negative");
+    return false;
+  }
+  if (padding > std::numeric_limits<double>::max())
+  {
+    ROS_ERROR_NAMED("collision_detection", "Padding must be finite");
+    return false;
+  }
+  return true;
+}
+
+namespace collision_detection
+{
+CollisionEnv::CollisionEnv(const robot_model::RobotModelConstPtr& model, double padding, double scale)
+  : robot_model_(model), world_(new World()), world_const_(world_)
+{
+  if (!validateScale(scale))
+    scale = 1.0;
+  if (!validatePadding(padding))
+    padding = 0.0;
+
+  const std::vector<const robot_model::LinkModel*>& links = robot_model_->getLinkModelsWithCollisionGeometry();
+  for (auto link : links)
+  {
+    link_padding_[link->getName()] = padding;
+    link_scale_[link->getName()] = scale;
+  }
+}
+
+CollisionEnv::CollisionEnv(const robot_model::RobotModelConstPtr& model, const WorldPtr& world, double padding, double scale)
+  : robot_model_(model), world_(world), world_const_(world_)
+{
+  if (!validateScale(scale))
+    scale = 1.0;
+  if (!validatePadding(padding))
+    padding = 0.0;
+
+  const std::vector<const robot_model::LinkModel*>& links = robot_model_->getLinkModelsWithCollisionGeometry();
+  for (auto link : links)
+  {
+    link_padding_[link->getName()] = padding;
+    link_scale_[link->getName()] = scale;
+  }
+}
+
+CollisionEnv::CollisionEnv(const CollisionEnv& other, const WorldPtr& world) : robot_model_(other.robot_model_), world_(world), world_const_(world)
+{
+  link_padding_ = other.link_padding_;
+  link_scale_ = other.link_scale_;
+}
+void CollisionEnv::setPadding(double padding)
+{
+  if (!validatePadding(padding))
+    return;
+  std::vector<std::string> u;
+  const std::vector<const robot_model::LinkModel*>& links = robot_model_->getLinkModelsWithCollisionGeometry();
+  for (auto link : links)
+  {
+    if (getLinkPadding(link->getName()) != padding)
+      u.push_back(link->getName());
+    link_padding_[link->getName()] = padding;
+  }
+  if (!u.empty())
+    updatedPaddingOrScaling(u);
+}
+
+void CollisionEnv::setScale(double scale)
+{
+  if (!validateScale(scale))
+    return;
+  std::vector<std::string> u;
+  const std::vector<const robot_model::LinkModel*>& links = robot_model_->getLinkModelsWithCollisionGeometry();
+  for (auto link : links)
+  {
+    if (getLinkScale(link->getName()) != scale)
+      u.push_back(link->getName());
+    link_scale_[link->getName()] = scale;
+  }
+  if (!u.empty())
+    updatedPaddingOrScaling(u);
+}
+
+void CollisionEnv::setLinkPadding(const std::string& link_name, double padding)
+{
+  bool update = getLinkPadding(link_name) != padding;
+  link_padding_[link_name] = padding;
```

Comment:
Can we add checks to `validatePadding` here?

Reply:
yes added it.